tuf/signed/sign.go

```diff
@@ -13,73 +13,76 @@ package signed
 
 import (
 	"crypto/rand"
-	"fmt"
 
 	"github.com/Sirupsen/logrus"
+	"github.com/docker/notary/trustmanager"
 	"github.com/docker/notary/tuf/data"
 	"github.com/docker/notary/tuf/utils"
 )
 
-// Sign takes a data.Signed and a key, calculated and adds the signature
-// to the data.Signed
-func Sign(service CryptoService, s *data.Signed, keys ...data.PublicKey) error {
-	logrus.Debugf("sign called with %d keys", len(keys))
-	signatures := make([]data.Signature, 0, len(s.Signatures)+1)
-	signingKeyIDs := make(map[string]struct{})
-	ids := make([]string, 0, len(keys))
+// Sign takes a data.Signed and replaces data.Signature with at least
+// minPrimarySignature signatures using primaryKeys and any possible
+// signatures using optionalKeys
+func Sign(service CryptoService, s *data.Signed, primaryKeys []data.PublicKey,
+	minSignatures int, optionalKeys []data.PublicKey) error {
+	logrus.Debugf("sign called with %d/%d required + %d optional keys",
+		minSignatures, len(primaryKeys), len(optionalKeys))
 
 	privKeys := make(map[string]data.PrivateKey)
 
 	// Get all the private key objects related to the public keys
-	for _, key := range keys {
+	missingKeyIDs := []string{}
+	for _, key := range primaryKeys {
 		canonicalID, err := utils.CanonicalKeyID(key)
-		ids = append(ids, canonicalID)
 		if err != nil {
-			continue
+			return err
 		}
 		k, _, err := service.GetPrivateKey(canonicalID)
 		if err != nil {
-			continue
+			if _, ok := err.(trustmanager.ErrKeyNotFound); ok {
+				missingKeyIDs = append(missingKeyIDs, canonicalID)
+				continue
+			}
+			return err
 		}
 		privKeys[key.ID()] = k
 	}
 
-	// Check to ensure we have at least one signing key
-	if len(privKeys) == 0 {
-		return ErrNoKeys{KeyIDs: ids}
+	// Check to ensure we have enough signing keys
+	if len(privKeys) < minSignatures {
```

Comment:
We are completely unprepared for true multi-person signing in any case: for that to work, there would have to be a mechanism for the first user to sign an updated file, and then to transfer the half-signed file to the other user, who would sign it again and then publish. We don’t have any file transfer mechanism other than the server (which refuses, or should refuse, half-signed files, and they should not be visible to the public anyway), and [incrementing the version = breaking old signatures, signing, and publishing] is linked into a single `NotaryRepository.Publish()` so there is not even a way to do this manually at the moment.

All of the code would have to be restructured to support thresholds anyway; this is at worst irrelevant to such a restructuring.

(Perhaps this is actually moving towards the easiest way to _implement_ thresholds: We could create a new `RemoteCryptoService` which transfers blobs for signing to the other user. That would allow us to implement multiple signatures without touching the `NotaryRepository.Publish` workflow at all — and in such a system, we do want to fail if we can not collect enough signatures, just like this PR is doing.)